[PT1.13 cherry pick]Fix Transformer's issue when padding is long type & unit test

zrphercule opened the pull request:
Summary:
Fix the issue described in
https://github.com/pytorch/pytorch/issues/86120

Test Plan: buck test mode/opt caffe2/test:test_transformers -- test_train_with_long_type_pad

Differential Revision: D40129968

ngimel commented on torch/nn/functional.py:
```diff
@@ -5071,6 +5071,10 @@ def multi_head_attention_forward(
         warnings.warn("Byte tensor for key_padding_mask in nn.MultiheadAttention is deprecated. Use bool tensor instead.")
         key_padding_mask = key_padding_mask.to(torch.bool)
 
+    if key_padding_mask is not None and key_padding_mask.dtype != torch.bool:
+        warnings.warn("Non-Boolean tensor for key_padding_mask in nn.MultiheadAttention is not recommended. Use bool tensor instead.")
```
To be honest, I'd prefer erroring out on mask types that are not bool or byte, because the authors intent is unclear in this case, and it's likely an error. For the tests that were failing after this change, what's the use case for long mask?

malfet replied:
This should be an error, rather than warning (even for byte type)

zrphercule committed:
[PT1.13 cherry pick]Fix Transformer's issue when padding is long type & unit test (#348)

Summary:
X-link: https://github.com/facebookresearch/multimodal/pull/348

Pull Request resolved: https://github.com/pytorch/pytorch/pull/86353

Fix the issue described in
https://github.com/pytorch/pytorch/issues/86120

Test Plan: buck test mode/opt caffe2/test:test_transformers -- test_train_with_long_type_pad

Differential Revision: D40129968

fbshipit-source-id: 3a6c6b4950889824a36f8d81b91cfe51fe9e9f5a

malfet commented:
Removing milestone for now, as PR is not yet passing the CI. @zrphercule any plans to fix the signal and land it?

malfet commented on torch/nn/functional.py:
```diff
@@ -5014,6 +5014,8 @@ def multi_head_attention_forward(
     # set up shape vars
     tgt_len, bsz, embed_dim = query.shape
     src_len, _, _ = key.shape
+    assert key_padding_mask is None or key_padding_mask.dtype == torch.bool or key_padding_mask.dtype == torch.float, \
```
Hmm, `assert` sounds like a bad mechanism for me, as they could be suppressed via a runtime settings.
It's better to raise an `AssertionError`

malfet reviewed (changes requested):
Please do not use `assert` but rather explicitly raise exception:
```
  if not cond:
     raise RuntimeError
```

zrphercule commented:
Seems phabricator's linking has some issue, will export a new one

mikaylagawarecki commented:
@zrphercule if this is BC-breaking and was cherry picked into 1.13 release could you please write a description for it, see 
[here](https://fb.quip.com/OCRoAbEvrRD9#HdaACAqenJa) for instructions. The release is soon so it would be great if you could do this ASAP. Thank you!
